Hotfix/remove add class from cmdu message

tomereli opened the pull request:
Prevent cases where parse is called more than once by making parse a CmduMessageTx method instead of CmduMessage. Also, make ClassList msg a protected member.

Commit:
tlvf: Move addClass wrapper to CmduMessageTx

Now that we parse by default, cmdu_rx.addClass should never be used on
parsed CMDUs, so move it to the CmduMessageTx where it should be used.
Also fix the following bugs found thanks to this change in places which
still did cmdu_rx.addClass() instead of cmdu_rx.getClass():

1. Fix print_cmdu_types to use getClass instead of addClass and add the
beerocks header as an inner classList so it will be swapped back when
the whole CMDU is swapped back
2. Backhaul manager thread use getClass instead of addClass in
ACTION_BACKHAUL_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION which got missed
when the automatic parser was added.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: CmduMessage make ClassList msg protected

To prevent accidental usage of ClassList::addClass for parsed CMDUs,
make the Classlist member protected instead of public.

Add a wrapper for getClassList<T> which was up to now accessed through
the ClassList member.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

CoralMalachi reviewed (commented):
I don't fully understand the commit message of the second commit.
Why did you say: "To prevent accidental usage of ClassList::getClass for parsed CMDUs", is not the `getClass` method is for parsed CMDUs ? or did you mean to say for not parsed CMDUs?

arnout commented on common/beerocks/tlvf/src/src/beerocks_message.cpp:
```diff
@@ -120,7 +120,7 @@ std::string message_com::print_cmdu_types(ieee1905_1::CmduMessageRx &cmdu_rx, sC
             return info;
         }
-
+        tlv_header->addInnerClassList(hdr); // so it will also be swapped back
```
This change isn't explained in the commit message.

arnout commented on framework/tlvf/src/include/tlvf/CmduMessage.h:
```diff
@@ -44,16 +42,24 @@ class CmduMessage {
     size_t getMessageLength() const { return msg.getMessageLength(); };
     size_t getMessageBuffLength() const { return msg.getMessageBuffLength(); };
     uint8_t *getMessageBuff() const { return msg.getMessageBuff(); };
+    template <class T> std::shared_ptr<T> getClass() { return msg.getClass<T>(); }
```
There is already a const getClass version a bit higher up, why do we also need a non-const one?

tomereli replied:
we don't, nice catch.

arnout reviewed (approved):
As noted by @CoralMalachi, the commit message of the second commit is bogus. Suggested improvement:

```
To prevent accidental usage of ClassList::addClass for parsed CMDUs,
make the ClassList member protected instead of public.

Add a wrapper for getClassList<T> which was up to now accesses through
the ClassList member.

```

Commit:
test_flows.sh: fix dummy events bugs

When simulating a dummy event, we update the
/tmp/$USER/beerocks/wlanX/EVENT file - the first line is always the one
that is parsed by the dummy BWL, so if we want to simulate an event we
must update the *first* line in this file.

Therefore, any update to the file which changes any other line will
result with actually simulating the last event instead of the new one.

Fix a couple of places in test_flow.sh which used `>>` which appends the
event to the end of the file and thus simulating the previous event and
replace it with `>` which overrides the file resulting with simulating
the correct and new event.

Fixed tests: client_association_dummy, client_steering_dummy.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
test_flows.sh: add --skip-init option

In some cases we want to debug certain tests after already running some
tests. For that, we don't want to kill the containers and restart
everything, so add --skip-init option to test_flow.sh which allows
running all tests again without restarting the containers.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
test_flows.sh: add --stop-on-failure option

On some occasions, we would like to stop on the first test failure
instead of iterating through all tests, for easier debugging so we don't
get a log of unrelated logs from other passing tests.
So add new option for test_flows.sh -s|--stop-on-failure which when set,
will stop at the first failed test.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bpl: linux: set stop_on_failure=1 in dummy mode

Setting stop_on_failure=0 results with endless slave reset on every
failure. We originally set this as the default mode in certification
since on dev_reset_default we expect a slave reset, but this is not
needed now in the new implementation since on dev_reset_default we
disconnect the slaves in the backhaul manager which results with
ACTION_BACKHAUL_DISCONNECTED_NOTIFICATION in the son_slave_thread. In
this case we do slave_reset() but do not decrement
stop_on_failure_attempts.

In addition, setting stop_on_failure=0 is really bad for development
since every slave reset can potentially hide a real bug if we fail
to realize that a slave reset occured.
It is especially difficult to spot this in CI.

Due to the above, there is no reason to set stop_on_failure=0 - so set
it to 1.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: dummy: fix empty events bug

BWL supports dummy event simulation by monitoring
/tmp/$USER/beerocks/wlanX/EVENT file - every time the file is written,
the BWL dummy implementation reads the first line and uses it as the
simulated event.

To simulate an event, we use 'echo foo > bar' format which first
generates an inotify event for the truncation, and then an event for the
new content. So, add a check and ignore empty events.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -544,10 +544,13 @@ test_topology() {
-        exit 1
+    status "test initialization"
+
+    [ "$1" = "false" ] && {
```
It would make more sense to use global SKIP_INIT here like we do for e.g. verbose. (nitpick)

tomereli replied:
Done.

arnout commented on framework/platform/bpl/linux/prplmesh_platform_db:
```diff
@@ -1,4 +1,4 @@
 management_mode=Multi-AP-Controller-and-Agent
 operating_mode=Gateway
-stop_on_failure_attempts=0
+stop_on_failure_attempts=1
```
I'm not convinced... We also want a test_flows for agent certification, and that will have to do dev_reset_default...

tomereli replied:
we don't decrement stop_on_failure_attempts when backhaul disconnects so setting stop_on_failure=1 is fine and will not hurt dev_reset_default.

See line 1323 in son_slave_thread.cpp - we call slave_reset() without decrementing stop on failure count so it is not counted as failure.

arnout reviewed (approved):
I think "bpl: linux: set stop_on_failure=1 in dummy mode" should be dropped. For debugging, you can just set it to 0 in the install dir before starting test_flows.sh.

tomereli commented:
> I think "bpl: linux: set stop_on_failure=1 in dummy mode" should be dropped. For debugging, you can just set it to 0 in the install dir before starting test_flows.sh.

See https://github.com/prplfoundation/prplMesh/pull/624#discussion_r362511205
